azure/packages/test/end-to-end-tests/src/test/signals.spec.ts

```diff
@@ -0,0 +1,230 @@
+		{ durationMs: timeoutMs, errorMsg: `${name}: listener timeout` },
+	);
+}
+
+describe("Fluid Signals", () => {
+	const connectTimeoutMs = 10_000;
+	const user1: AzureUser = {
+		id: "test-user-id-1",
+		name: "test-user-name-2",
+	};
+	const user2: AzureUser = {
+		id: "test-user-id-1",
+		name: "test-user-name-2",
+	};
+	const user3: AzureUser = {
+		id: "test-user-id-1",
+		name: "test-user-name-2",
+	};
+
+	const getOrCreateSignalerContainer = async (
+		id: string | undefined,
+		user: AzureUser,
+		config?: ReturnType<typeof configProvider>,
+		scopes?: ScopeType[],
+	): Promise<{
+		container: IFluidContainer;
+		signaler: Signaler;
+		services: AzureContainerServices;
+		client: AzureClient;
+		containerId: string;
+	}> => {
+		const client = createAzureClient(user.id, user.name, undefined, config, scopes);
+		const schema: ContainerSchema = {
+			initialObjects: {
+				signaler: Signaler,
+			},
+		};
+		let container: IFluidContainer;
+		let services: AzureContainerServices;
+		let containerId: string;
+		if (id === undefined) {
+			({ container, services } = await client.createContainer(schema));
+			containerId = await container.attach();
+		} else {
+			containerId = id;
+			({ container, services } = await client.getContainer(containerId, schema));
+		}
+
+		if (container.connectionState !== ConnectionState.Connected) {
+			await timeoutPromise((resolve) => container?.once("connected", () => resolve()), {
+				durationMs: connectTimeoutMs,
+				errorMsg: "container connect() timeout",
+			});
+		}
+
+		assert.strictEqual(typeof containerId, "string", "Attach did not return a string ID");
+		assert.strictEqual(
+			container.attachState,
+			AttachState.Attached,
+			"Container is not attached after attach is called",
+		);
+
+		const signaler = container.initialObjects.signaler as Signaler;
+		return {
+			client,
+			container,
+			signaler,
+			services,
+			containerId,
+		};
+	};
+
+	/**
+	 * Scenario: Client sends a signal and connected clients receive it.
+	 *
+	 * Expected behavior: While 2 clients are connected to a container,
+	 * a signal sent by 1 client should be recieved by both clients.
+	 */
+	it("can send and receive signals", async () => {
+		const { signaler, containerId } = await getOrCreateSignalerContainer(undefined, user1);
+		const { signaler: signaler2 } = await getOrCreateSignalerContainer(
+			containerId,
+			user2,
+			configProvider({
+				"Fluid.Container.ForceWriteConnection": true,
+			}),
+		);
+
+		const signalName = "test-signal";
+		const signalPayload = { test: "payload" };
+
+		const listenerPromises = [
+			createSignalListenerPromise(
+				signaler2,
+				signalName,
+				signalPayload,
+				"Write client listening for write client signal",
+			),
+			createSignalListenerPromise(
+				signaler,
+				signalName,
+				signalPayload,
+				"Write client listening for its own signal",
+			),
+		];
+
+		signaler.submitSignal(signalName, signalPayload);
+
+		await Promise.all(listenerPromises);
```

Comment:
Are you trying to see if `await Promise.all(listenerPromises);` throws? If so, have you considered using `assert.doesNotThrow()`?

Reply:
Good idea, thanks!